File: python_backend/fenParser/Fen.py
from custom_except import *
from coordType.to_rankfile import to_rankfile


class Fen:
    def __init__(self, fen, turn, castle_avail, en_passant_avail, hm_clock, fm_clock):
        self.fen = fen
        self.turn = turn
        self.castle_avail = castle_avail
        self.en_passant_avail = en_passant_avail
        self.hm_clock = int(hm_clock)
        self.fm_clock = int(fm_clock)

    def update_state(self, special_moves, json_records, start, dest, captured, color):
        """update the non piece-position attributes of the fen: turn, castle avail, en-passant avail, and clocks"""
        self.turn = color.lower()
        self._update_castle_avail(json_records)
        self.update_en_passant_avail(dest, special_moves, start)
        self._update_clocks(captured, color, json_records)

    def _update_clocks(self, captured, color, json_records):
        if json_records.num_consecutive_non_pawn_moves == 0 or captured != 'None':
            self.hm_clock = 0
        if color == 'B':
            self.fm_clock += 1

    def update_en_passant_avail(self, dest, special_moves, start):
        if (start, dest) in special_moves.en_passant:
            self.en_passant_avail = to_rankfile(dest)
        else:
            self.en_passant_avail = '-'

    def _update_castle_avail(self, json_records):
        if not json_records.kings_moved[(5, 1)] and not json_records.rooks_moved[(8, 1)]:
            K = 'K'
        else:
            K = ''
        if not json_records.kings_moved[(5, 1)] and not json_records.rooks_moved[(1, 1)]:
            Q = 'Q'
        else:
            Q = ''
        if not json_records.kings_moved[(5, 8)] and not json_records.rooks_moved[(8, 8)]:
            k = 'k'
        else:
            k = ''
        if not json_records.kings_moved[(5, 8)] and not json_records.rooks_moved[(8, 1)]:
            q = 'q'
        else:
            q = ''
        self.castle_avail = K + Q + k + q
        if self.castle_avail == '':
            self.castle_avail = '-'

    def get_data(self):
        """return the fen components as in an object that can be sent to React"""
        return {"fen": self.fen, "turn": self.turn, "castle_avail": self.castle_avail,
                "en_passant_avail": self.en_passant_avail,
                "hm_clock": self.hm_clock, "fm_clock": self.fm_clock}

    def set_get_new_fen(self, pos):
        """join the position string with the status attributes updated throughout the game, then return"""
        self.fen = " ".join(
            [pos, self.turn, self.castle_avail, self.en_passant_avail, str(self.hm_clock), str(self.fm_clock)])
        return self.fen

File: app/src/game_logic/fenParser/Fen.js
import {strfind} from "../helpers/strfind";


export class Fen {

    constructor(data) {
        this.fen = data["fen"]
        this.turn = data["turn"]
        this.castleAvail = data["castle_avail"]
        this.enPassantAvail = data["en_passant_avail"]
        this.hmClock = parseInt(data["hm_clock"], 10)
        this.fmClock = parseInt(data["fm_clock"], 10)
    }

    update(specialMoves, jsonRecords, start, dest, captured, color) {
        /*update the non piece-position attributes of the fen: turn, castle avail, en-passant avail, &&  clocks**/
        this.turn = color.toLowerCase()
        this.UpdateCastleAvail(jsonRecords)
        this.updateEnPassantAvail(dest, specialMoves, start)
        this.UpdateClocks(captured, color, jsonRecords)
    }

    UpdateClocks(captured, color, jsonRecords) {
        if (jsonRecords.numConsecutiveNonPawnMoves === 0 || captured != 'undefined') {
            this.hmClock = 0
        }
        if (color === 'B') {
            this.fmClock += 1
        }
    }

    updateEnPassantAvail(dest, specialMoves, start) {
        if (strfind(specialMoves.enPassant, [start, dest])) { //condition needs modification
            this.enPassantAvail = dest;
        }
        else {
            this.enPassantAvail = '-'
        }
    }

    UpdateCastleAvail(jsonRecords) {

        var K, Q, k, q;

        if (! jsonRecords.kingsMoved["e1"] && ! jsonRecords.rooksMoved["h1"]) {
            K = 'K'
        }
        else {
            K = ''
        }
        if (! jsonRecords.kingsMoved["e1"] && ! jsonRecords.rooksMoved["a1"]) {
            Q = 'Q'
        }
        else {
            Q = ''
        }
        if (! jsonRecords.kingsMoved["e8"] && ! jsonRecords.rooksMoved["a8"]) {
            k = 'k'
        }
        else {
            k = ''
        }
        if (! jsonRecords.kingsMoved["e8"] && ! jsonRecords.rooksMoved["h8"]) {
            q = 'q'
        }
        else {
            q = ''
        }

        this.castleAvail = K + Q + k + q
        if (this.castleAvail === '') {
            this.castleAvail = '-'
        }

    }

    getData() {
        /*return the fen components an.includes(as) object that can be sent to React**/
        return {"fen": this.fen, "turn": this.turn, "castle_avail": this.castleAvail,
                "en_passant_avail": this.enPassantAvail,
                "hm_clock": this.hmClock, "fm_clock": this.fmClock}
    }

    setGetNewFen(pos) {
        /*join the position string with the status attributes updated throughout the game, then return**/
        this.fen = " ".join([pos, this.turn, this.castleAvail, this.enPassantAvail, this.hmClock.toSring(), this.fmClock.toString() ]);
        return this.fen;
    }
}

// module.exports = Fen;